R versions

trws opened the pull request:
Fix issues discussed in #29250.

trws committed:
lower priority of package-provided urls

This change favors urls found in a scraped page over those provided by
the package from `url_for_version`.  In most cases this doesn't matter,
but R specifically returns known bad URLs in some cases, and the
fallback path for a failed fetch uses `fetch_remote_versions` to find a
substitute.  This fixes that problem.

fixes #29204

trws committed:
consider what links actually exist in all cases

Checksum was only actually scraping when called with no versions.  It
now always scrapes and then selects URLs from the set of URLs known to
exist whenever possible.

fixes #25831

trws committed:
bow to the wrath of flake8

glennpj commented:
I can confirm that this fixes the issues discussed in #29250, including #29204 and #26977, which could be closed by this. I checked

1.  fetching the version set equal to that in the url, when that archive has been moved on the remote and needs `list_url`
2.  checksum without a specific version on the command line
3. checksum with a specific version on the command line. 

All seems to work now.

sethrj commented on lib/spack/spack/cmd/checksum.py:
```diff
@@ -56,33 +56,29 @@ def checksum(parser, args):
 
-    url_dict = {}
+    # Otherwise, see what versions we can find online
+    url_dict = pkg.fetch_remote_versions()
```
Before this was done only conditionally (no args given) because it can be extremely expensive (e.g. `qt` which does a lot of spidering). What's the runtime of `spack checksum qt@5.nonexistent` on this branch?

trws replied:
It is expensive, but it's also the only way to get correct behavior for packages where the URLs generated by their packages may or may not be correct.  The direct answer to your question is that `qt@5.nonexistent` takes 3 minutes and 43 seconds.

As far as I can think right now, the only way to deal with this better is to make it illegal for a package to generate broken links, but then R packages can't work.  We could work around that by changing the contract for generating URLs from a version to produce a list of URLs, at least one of which must be valid if the version is valid, but that would be a breaking change.

I suppose one intermediate step might be that we can do a check on the provided url and see if we get a 404, and only then fetch remote versions if we hit a failure from the package generated url?

sethrj replied:
Yuck. Does it take that long to start downloading a valid version, such as 5.15.2? I know the qt "list URLs" is a bit of an outlier but it's going to be ugly if a new version comes out but takes minutes to verify the URL, even if we know it exists.

trws replied:
For `spack checksum` yes it does, for `spack fetch` it does not (`spack fetch qt@5.15.2` takes 16 seconds total with download time), this only gets called if the first URL fetch tries fails.  We could do basically the same thing for checksum to shortcut it for valid URLs, it's not quite that simple but it is possible.

sethrj replied:
Yes, it would be ideal if you could make `pkg.fetch_remote_versions` a lazy evaluation... it's something of a nuclear option, and because it uses the URL spidering can be unstable -- best if the checksum command only calls it if really necessary.

trws replied:
Ok, let me prototype something.  It will have to be a bigger change, and there will still be unconditional requests, but cheaper ones that don't involve spidering by default.

sethrj replied:
Thanks @trws !

glennpj replied:
> As far as I can think right now, the only way to deal with this better is to make it illegal for a package to generate broken links, but then R packages can't work.

The R package locations are actually very predictable. The spidering should not be too bad as each package has a directory under "Archive", the `list_url` points to it, and there should not be that much in those respective directories. The location of an older source package is actually known but there is no mechanism in spack to fetch it other than `list_url`, at least none that I am aware of.

This may be out of scope for this PR but would it be feasible to have a second `url` attribute in spack? An R package could then look something like this:
```
url = "https://cloud.r-project.org/src/contrib/abind_1.4-5.tar.gz"
archive_url = "https://cloud.r-project.org/src/contrib/Archive/abind/abind_1.4-5.tar.gz"
```

The above would be handled by the `cran` attribute as `url`, `list_url`, and `homepage` are currently. Basically, if the source is not found at `url` try `archive_url`, using the same logic for replacing version numbers.

trws replied:
Ok, so actually it should use the `urls` attribute rather than the `url` attribute, that will let both be specified and both will be tried.  The way it is now, with one as `list_url` will also work after this change.

trws replied:
To be more complete about it, the code used to assume that the package could always provide a single url that could be determined to be valid without doing a test fetch.  That doesn't really work in the general case, this makes the fallback cases a whole lot more explicit, and uses them in more places.  The fetch strategy already did some of this, but checksum and similar did not.

glennpj replied:
The `urls` attribute currently fails for fetching the older versions of packages listed in the package file. Interestingly, `checksum` works for any versions **not** listed in the package file. Is that something you are trying to fix here as well?

glennpj replied:
Sorry, I should have pulled the latest changes before posting. With the most recent changes in this PR, `spack checksum` now works with the `urls` attribute. Fetching/staging still fails though.

trws replied:
Ok now that's interesting.  The change I had but walked back would fix that, I expected current behavior would as well.  I'm tempted to say keep it list_url for now, and we can open something to discuss whether the fetch strategy should use my new logic for using fetch checks by default.  Honestly I think it should, but there's a unit test that walks every version in every package, and generates fetch strategies for them.  That is a recipe for disaster so I walked it back to use just the new url listing stuff, but it should check all the mirrors rather than just dying... that seems like a separate issue though since it dies currently too I guess.

trws committed:
test-fetch urls from package, prefer if successful

trws commented:
As far as I can tell, this version now solves the issues, and avoids massive performance regression on heavy spidering packages.  @sethrj, @glennpj, you good with this?

sethrj commented on lib/spack/spack/cmd/checksum.py:
```diff
@@ -56,33 +56,35 @@ def checksum(parser, args):
+            # if we get here, it's because no valid url was provided by the package
+            # do expensive fallback to try to recover
+            if url_dict is None:
+                url_dict = pkg.fetch_remote_versions()
```
I think this code would be much easier to follow if you modified `url_dict` instead of `new_dict`, and define `remote_versions = None` inside the `if versions` branch that holds the lazily evaluated remote versions.

trws replied:
That's fair, update incoming.

sethrj commented on lib/spack/spack/package.py:
```diff
@@ -2561,6 +2634,7 @@ def fetch_remote_versions(self, concurrency=128):
         if not self.all_urls:
             return {}
 
+        tty.debug("in fetch_remote_versions")
```
debug code left in

trws replied:
Thanks for that. 
```suggestion
```

trws committed:
Update lib/spack/spack/package.py

Co-authored-by: Seth R. Johnson <johnsonsr@ornl.gov>

trws committed:
reword as suggested

adamjstewart commented on lib/spack/spack/package.py:
```diff
@@ -1298,7 +1371,7 @@ def prefix(self):
         return self.spec.prefix
 
-    @property  # type: ignore[misc]
+    @property  # type: ignore
```
We should only ignore the specific mypy error instead of blanket ignoring all

trws replied:
I removed that because ignore[misc] now causes flake8 to fail: 

```
==> Running flake8 checks
  lib/spack/spack/package.py:1374: [F821] undefined name 'ignore'
  lib/spack/spack/package.py:1374: [F821] undefined name 'misc'
```

adamjstewart replied:
Hmm, I use mypy + flake8 in my projects and I've never had them conflict like this before...

trws replied:
Added it back, and ignored the error from pyflakes.

trws committed:
re-enable mypy specific ignore and ignore pyflakes

adamjstewart commented on .flake8:
```diff
@@ -35,8 +35,10 @@
 # - E203: white space around slice operators can be required, ignore : warn
 # - W503: see above, already ignored for line-breaks
 #
+# F821: generates an error on mypy comments with error codes like # ignore[misc]
```
We don't want to ignore this globally since it means all undefined variables will be completely ignored.

It looks like this was a bug in older versions of pyflakes (https://github.com/PyCQA/pyflakes/issues/475), fixed in 2.2.0 (https://github.com/PyCQA/pyflakes/pull/455). Can you try a newer version of pyflakes/flake8 instead?

trws replied:
Reinstated, and verified with a pipx installed copy on my machine. Thanks for the catch.

trws committed:
remove flake8 ignore from .flake8

scheibelp commented on lib/spack/spack/package.py:
```diff
@@ -973,6 +973,79 @@ def url_for_version(self, version):
         return spack.url.substitute_version(
             default_url, self.url_version(version))
 
+    def urls_for_version(self, version):
```
If this were a generator, could `url_for_version` just pick the first value yielded? I'm noticing some logical overlap between this function and `url_for_version`, and I also find it confusing to have functions named `url_for_version` and `urls_for_version`: perhaps if this were called `all_urls_for_version` that would help distinguish it from `url_for_version`.

trws replied:
Probably, or at least I could make that possible.

trws replied:
This turns out to be harder than I would have thought.  Since `url_for_version` is the customization point for packages, I need to call it in `all_urls_for_version`.  I tried to detect when it's customized and only call it then, but that is proving to be unreliable.  I'll be happy to do the rename, but getting the singular to use the multiple is not straightforward.

scheibelp commented on lib/spack/spack/cmd/checksum.py:
```diff
@@ -56,33 +56,34 @@ def checksum(parser, args):
     # Get the package we're going to generate checksums for
     pkg = spack.repo.get(args.package)
 
+    # Otherwise, see what versions we can find online
```
I think this comment no longer makes sense here (the online search was moved to later)

scheibelp commented on lib/spack/spack/cmd/checksum.py:
```diff
@@ -56,33 +56,34 @@ def checksum(parser, args):
+            if remote_versions is None:
+                remote_versions = pkg.fetch_remote_versions()
+            if version in remote_versions:
+                url_dict[version] = remote_versions[version]
```
(question) this looks like a preexisting problem but to make sure: if this `if` check fails, it looks like we should raise an exception here: at this point there is no URL available to fetch the version and checksum it.

trws replied:
Right, the current behavior is it only fails if none of the versions are found, rather than if any of them are not found.

scheibelp commented on lib/spack/spack/test/url_fetch.py:
```diff
@@ -40,8 +48,16 @@ def fn(v):
 
+        def fn_urls(v):
+            urls_loc = urls or [url]
+            return [spack.url.substitute_version(main_url, v) for main_url in urls_loc]
```
(minor) IMO it's confusing to use `main_url` as the iteration variable name since I thought only the first URL was the "main" one. In other words: what does "main" mean here?

trws replied:
It means "I copy and pasted it," I'll make it `u` or something.

scheibelp commented on lib/spack/spack/util/web.py:
```diff
@@ -665,6 +663,11 @@ def find_versions_of_archive(
+    for url in archive_urls:
+        ver = spack.url.parse_version(url)
+        if ver not in versions:
+            versions[ver] = url
```
(question) I think placing this logic here inverts the original prioritization - is that intended?

i.e. originally the `archive_urls` would take precedence (because of `if ver in matched: continue`), but now they are only used if the links are not.

trws replied:
Yes, it's very much intentional, that's to fix one of the explicit bugs.  The issue is that sometimes the URLs from the package aren't known to be correct, so prioritizing the archive_urls was pinning non-functional links without a way to tell.  Now it prioritizes a detected link over the archive URL on the assumption that if it found it that way it's more likely to be valid.  I would worry about this causing issues if this function were always used first, but it isn't, it's normally used as a fallback after the archive URL or URLs have been tried.

scheibelp replied:
That sounds good: I think it would be worthwhile to put this explanation in the docstring (or rather to explain what gets the highest priority, not necessarily to delve into why).

scheibelp reviewed (changes requested):
I have a couple questions and a couple of requests.

trws committed:
address review comments

trws commented:
Ok, I think that takes care of your issues @scheibelp.  The one thing that feels
unfinished here is I feel like I ought to be using `find_valid_url_for_version`
in `_from_merged_attrs` since the fetchers do not seem to actually try all the
mirrors for some reason and it un-breaks certain kinds of broken package links.
I backed out that change because there's a unit-test that would test-fetch every
single package version in all of spack with that change in there.  I still think
it would be a good idea to do it, but it clearly violates an assumption in terms
of expense to oo it right now, so I'm leaving that for later.

trws commented:
Any chance of a re-review @scheibelp?

scheibelp commented on lib/spack/spack/package.py:
```diff
@@ -973,6 +973,80 @@ def url_for_version(self, version):
+
+        version: class Version
+            The version for which a URL is sought.
+        use_list_url: bool
```
This parameter does not exist

trws replied:
Ah, thanks for that, forgot to remove it in a refactor.

scheibelp commented on lib/spack/spack/package.py:
```diff
@@ -973,6 +973,80 @@ def url_for_version(self, version):
+            urls.append(version_urls[version])
+
+        # if there is a custom url_for_version, use it
+        u = self.url_for_version(version)
```
Regarding:

> This turns out to be harder than I would have thought. Since url_for_version is the customization point for packages, I need to call it in all_urls_for_version. I tried to detect when it's customized and only call it then, but that is proving to be unreliable. I'll be happy to do the rename, but getting the singular to use the multiple is not straightforward.

We could:

* Move all of this logic to a new function
* Add a `_custom_url_function` parameter to that function
* `all_urls_for_version` sets `_custom_url_function=url_for_version`
*  `url_for_version` can set `_custom_url_function=None`

trws replied:
Assuming my updated test for overloading, this should be in and working now.

trws committed:
address comments

scheibelp commented:
Thanks for the edits! FYI there appears to be a unit test failing. Let me know if that implies that my suggestion in https://github.com/spack/spack/pull/29258#discussion_r824193898 wasn't good - my goal is to avoid duplicate code though so I'm hoping it will work.

trws committed:
add sneaky missing substitute

I missed this one because we call substitute on a URL that doesn't
contain a version component.  I'm not sure how that's supposed to work,
but apparently it's required by at least one mock package, so back in it
goes.

trws commented:
@scheibelp it looks like the suggestion was just fine, I had missed a substitution in a case where the url doesn't have a version to replace, which seems interesting, but apprently it's required.  Aside from the codecov light it's looking green.

scheibelp commented on lib/spack/spack/util/web.py:
```diff
@@ -643,17 +647,15 @@ def find_versions_of_archive(
     versions = {}
     matched = set()
-    for url in archive_urls + sorted(links):
+    for url in sorted(links):
```
Hi @trws, apologies but there's a conflict here after merging https://github.com/spack/spack/pull/27021. This can be resolved by using your changes but adding `url = convert_to_posix_path(url)` at the beginning of this loop (and you will want to do the same down at line 670 i.e. `for url in archive_urls:`).

trws replied:
Ok, I think I got that resolved.  Here's hoping the tests bear that out. 🤞

trws committed:
Merge branch 'develop' into r-versions

trws committed:
Merge branch 'develop' into r-versions

trws committed:
Merge branch 'develop' into r-versions
